### PR TITLE
feat(account invitation): accept the invitation by the current user

### PR DESCRIPTION
**Pull request by andreybakanovsky**

### CODE REVIEW

File: app/views/account_invitations/accept_invitation.html.slim [hunk 1]
```diff
@@ -10,4 +10,5 @@ div.columns.is-centered.is-vcentered
       = link_to 'Accept', accept_account_invitation_path, method: :post, class: 'button is-success is-medium'
       span.has-text-grey-light
 
-      = link_to 'Cancel', root_path, class: 'button is-light is-medium'
\ No newline at end of file
+      = link_to 'Cancel', root_path, class: 'button is-light is-medium'
+
```
Comment (ka8725) [inline app/views/account_invitations/accept_invitation.html.slim:14]: fix this?

File: app/controllers/account_invitations_controller.rb [hunk 2]
```diff
@@ -18,8 +21,33 @@ def destroy
     end
   end
 
+  def accept_invitation
+    if params.key?(:invitation_id)
+      redirect_to root_path unless received_invitation && token_valid? && invitation_belongs_to_current_user?
+    else
```
Comment (ka8725) [inline app/controllers/account_invitations_controller.rb:27]: if no invitation_id key this action should not be activated, so remove this `if`?

File: app/controllers/account_invitations_controller.rb [hunk 3]
```diff
@@ -18,8 +21,33 @@ def destroy
     end
   end
 
+  def accept_invitation
```
Comment (ka8725) [inline app/controllers/account_invitations_controller.rb:24]:
```suggestion
  def review
```

File: config/routes.rb [hunk 4]
```diff
@@ -12,6 +12,10 @@
     resources :objectives, only: %i[new create]
     resources :invitations, only: %i[index new create destroy], controller: 'account_invitations'
   end
+
+  get 'account_invitation/:invitation_id', to: 'account_invitations#accept_invitation', as: :visit_by_account_invitation
+  post 'account_invitation/:invitation_id', to: 'account_invitations#accept', as: :accept_account_invitation
```
Comment (ka8725) [inline config/routes.rb:17]: can we put them into the block that defines invitations as members (https://guides.rubyonrails.org/routing.html#adding-member-routes)?

File: app/controllers/account_invitations_controller.rb [hunk 5]
```diff
@@ -18,8 +21,33 @@ def destroy
...
+    if params.key?(:invitation_id)
+      redirect_to root_path unless received_invitation && token_valid? && invitation_belongs_to_current_user?
```
Comment (ka8725) [inline app/controllers/account_invitations_controller.rb:26]: if `invitation_belongs_to_current_user?` then it's an error (no weird redirects), 404 would be fine.

also check that it was not used before. both an invalid or used invitation should give 404 error.

File: app/controllers/account_invitations_controller.rb [hunk 6]
```diff
@@ -18,8 +21,33 @@ def destroy
...
+  helper_method memoize def received_invitation
+    AccountInvitation.find(ps.fetch(:invitation_id))
+  end
+
+  def invitation_belongs_to_current_user?
+    current_user&.email == received_invitation.email
+  end
+
+  def token_valid?
+    received_invitation.token == ps.fetch(:token)
```
Comment (ka8725) [inline app/controllers/account_invitations_controller.rb:48]: In case of review/accept invitation the route should be like this:

`accept_account_invitation/:token`

Then we would not have this weird check for param. These two controllers would have significantly different implementations. Should we create two controllers? The second, for accept, can be named `AcceptAccountInvitationsController` with two standard methods `show` and `update`.

**Review by ka8725 — CHANGES_REQUESTED** [review]

**Comment by VladislavSokov** [comment]
I think this can be disabled in the linter

.rubocop.yml

Style/Documentation:
  Description: Document classes and non-namespace modules.
  Enabled: false

File: app/controllers/accept_account_invitations_controller.rb [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+  def update
+    received_invitation.update!(accepted_at: Time.now)
+    redirect_to account_path received_invitation.account_id
+  end
+
+  private
+
+  helper_method memoize def received_invitation
+    AccountInvitation.find_by(token: ps.fetch(:token))
```
Comment (ka8725) [inline app/controllers/accept_account_invitations_controller.rb:16]:
```suggestion
    AccountInvitation.unaccepted.for(current_user).find_by!(token: ps.fetch(:token))
```

File: app/controllers/accept_account_invitations_controller.rb [hunk 8]
```diff
@@ -0,0 +1,30 @@
+class AcceptAccountInvitationsController < ApplicationController
+  def show
+    not_found unless received_invitation &&
+                     invitation_belongs_to_current_user? &&
+                     not_accepted_invitation?
```
Comment (ka8725) [inline app/controllers/accept_account_invitations_controller.rb:5]: if add the scope `unaccepted` below and use `find_by!` instead of `find_by`, no need this check.

Comment (ka8725) [inline app/controllers/accept_account_invitations_controller.rb:5]:
```suggestion
```

follow the recommendations below and make it shorter like this ^

File: app/controllers/accept_account_invitations_controller.rb [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+  end
+
+  def update
+    received_invitation.update!(accepted_at: Time.now)
```
Comment (ka8725) [inline app/controllers/accept_account_invitations_controller.rb:9]:
```suggestion
    received_invitation.update!(accepted_at: Time.current)
```

Use current instead of now please as current takes time zone into account

File: app/controllers/accept_account_invitations_controller.rb [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+    redirect_to account_path received_invitation.account_id
```
Comment (ka8725) [inline app/controllers/accept_account_invitations_controller.rb:10]: use `()` - I'm having trouble reading it, is it `redirect_to account_path(received_invitation.account_id)` or `redirect_to(account_path received_invitation.account_id)`?

File: app/controllers/accept_account_invitations_controller.rb [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+    current_user&.email == received_invitation.email
+  end
+
+  def token_valid?
+    received_invitation.token == ps.fetch(:token)
+  end
+
+  def not_accepted_invitation?
+    received_invitation.accepted_at.nil?
+  end
```
Comment (ka8725) [inline app/controllers/accept_account_invitations_controller.rb:29]:
```suggestion
```

File: app/controllers/account_invitations_controller.rb [hunk 12]
```diff
@@ -1,12 +1,15 @@
 class AccountInvitationsController < ApplicationController
-  before_action :authenticate_user!
+  before_action :authenticate_user!, except: :accept_invitation
```
Comment (ka8725) [inline app/controllers/account_invitations_controller.rb:2]:
```suggestion
  before_action :authenticate_user!

File: app/controllers/accept_account_invitations_controller.rb [hunk 13]
```diff
@@ -0,0 +1,30 @@
+class AcceptAccountInvitationsController < ApplicationController
```
Comment (ka8725) [inline app/controllers/accept_account_invitations_controller.rb:1]: `before_action :authenticate_user!`?

File: app/controllers/application_controller.rb [hunk 14]
```diff
@@ -6,6 +6,10 @@ def ps
     request.params
   end
 
+  def not_found
+    render file: "#{Rails.root}/public/404.html", layout: false, status: :not_found
+  end
+
```
Comment (ka8725) [inline app/controllers/application_controller.rb:12]:
```suggestion
```

**Review by ka8725 — CHANGES_REQUESTED** [review]

File: app/controllers/accept_account_invitations_controller.rb [hunk 15]
```diff
@@ -0,0 +1,20 @@
+# frozen_string_literal: true
+
+class AcceptAccountInvitationsController < ApplicationController
+  before_action :authenticate_user!
+
+  def show
+    received_invitation
```
Comment (ka8725) [inline app/controllers/accept_account_invitations_controller.rb:7]:
```suggestion
```

it will be called from the UI when needed.

File: app/controllers/my_accounts_controller.rb [hunk 16]
```diff
@@ -1,11 +1,16 @@
 class MyAccountsController < ApplicationController
   before_action :authenticate_user!
 
-  def show
-  end
+  def show; end
 
   private
 
+  helper_method def accounts
```
Comment (ka8725) [inline app/controllers/my_accounts_controller.rb:8]: Remove this for now as it's not related to the current task.

File: app/models/account.rb [hunk 17]
```diff
@@ -15,7 +15,14 @@ class Account < ApplicationRecord
...
   scope :visible_for, lambda { |current_user|
-    where(id: [current_user.account_id] + current_user.account.child_ids)
+    where(id: [current_user.account_id] +
+               current_user.account.child_ids +
+               AccountInvitation.accepted_by(current_user).pluck(:account_id))
+  }
+
+  scope :children_and_invitees, lambda { |current_user|
+    where(id: current_user.account.child_ids +
+              AccountInvitation.accepted_by(current_user).pluck(:account_id))
```
Comment (ka8725) [inline app/models/account.rb:25]: This is not related to the current task. The current task is only about accept invitation, nothing more.

File: app/models/account.rb [hunk 18]
```diff
@@ -15,7 +15,14 @@ class Account < ApplicationRecord
   validates :name, presence: true
 
   scope :visible_for, lambda { |current_user|
-    where(id: [current_user.account_id] + current_user.account.child_ids)
+    where(id: [current_user.account_id] +
+               current_user.account.child_ids +
+               AccountInvitation.accepted_by(current_user).pluck(:account_id))
```
Comment (ka8725) [inline app/models/account.rb:20]: revert as it's not related to the current task

File: app/models/account_invitation.rb [hunk 19]
```diff
@@ -1,4 +1,11 @@
 class AccountInvitation < ApplicationRecord
   belongs_to :user
   belongs_to :account
+
+  scope :accepted_by, lambda { |current_user|
+                        where(email: current_user.email).where.not(accepted_at: nil)
+                      }
+  scope :unaccepted_for, lambda { |current_user|
+                           where(email: current_user.email).where(accepted_at: nil)
+                         }
```
Comment (ka8725) [inline app/models/account_invitation.rb:10]:
```suggestion

  scope :for, -> (user) { |user| where(email: user.email) }
  scope :accepted,  -> { where.not(accepted_at: nil) }
  scope :unaccepted, -> { where(accepted_at: nil) }
```

Separate scopes give more flexibility and can be combined in different combinations.

**Review by ka8725 — CHANGES_REQUESTED** [review]
Extract not related to the current task code. You can make a separate PR so as not to have this code lost. It seems correct, but it's hard to review it in this PR.

File: app/views/my_accounts/show.html.erb [hunk 20]
```diff
@@ -2,19 +2,22 @@
 <br>
 Account balance: <%= account.balance %>
 <br>
-<% account.children.each do |child| %>
-  <div class="card box">
+<% accounts.each do |account| %>
```
Comment (ka8725) [inline app/views/my_accounts/show.html.erb:5]: should be reverted too

**Review by ka8725 — APPROVED** [review]

File: spec/factories/account_invitations.rb [hunk 21]
```diff
@@ -2,8 +2,8 @@
   factory :account_invitation do
     association :user, factory: :user
     name { FFaker::Name.first_name }
-    email { FFaker::Internet.email }
-    token { FFaker::Lorem.phrase }
+    email { user.email }
```
Comment (ka8725) [inline spec/factories/account_invitations.rb:5]: Is it an invitation to themselves? Why changed it?

```suggestion
    email { FFaker::Internet.email }
```

**Review by VladislavSokov — APPROVED** [review]
